labelbox/data/annotation_types/data/tiled_image.py

```diff
@@ -0,0 +1,525 @@
+        """
+
+        if multithread:
+            tiles = {}
+            with ThreadPoolExecutor(
+                    max_workers=TILE_DOWNLOAD_CONCURRENCY) as exc:
+                for x in range(x_tile_start, x_tile_end + 1):
+                    for y in range(y_tile_start, y_tile_end + 1):
+                        tiles[(x, y)] = exc.submit(self._fetch_tile, x, y, zoom)
+
+        rows = []
+        for y in range(y_tile_start, y_tile_end + 1):
+            row = []
+            for x in range(x_tile_start, x_tile_end + 1):
+                try:
+                    if multithread:
+                        row.append(tiles[(x, y)].result())
+                    else:
+                        row.append(self._fetch_tile(x, y, zoom))
+                except:
+                    row.append(
+                        np.zeros(shape=(self.tile_size, self.tile_size, 3),
+                                 dtype=np.uint8))
+            rows.append(np.hstack(row))
+
+        return np.vstack(rows)
+
+    @retry.Retry(initial=1, maximum=16, multiplier=2)
+    def _fetch_tile(self, x: int, y: int, z: int) -> np.ndarray:
+        """
+        Fetches the image and returns an np array.
+        """
+        data = requests.get(self.tile_layer.url.format(x=x, y=y, z=z))
+        data.raise_for_status()
+        decoded = np.array(Image.open(BytesIO(data.content)))[..., :3]
+        if decoded.shape[:2] != (self.tile_size, self.tile_size):
+            logger.warning(f"Unexpected tile size {decoded.shape}.")
+        return decoded
+
+    def _crop_to_bounds(
+        self,
+        image: np.ndarray,
+        x_px_start: int,
+        y_px_start: int,
+        x_px_end: int,
+        y_px_end: int,
+    ) -> np.ndarray:
+        """This function slices off the excess pixels that are outside of the bounds.
+        This occurs because only full tiles can be downloaded at a time.
+        """
+
+        def invert_point(pt):
+            # Must have at least 1 pixel for stability.
+            pt = max(pt, 1)
+            # All pixel points are relative to a single tile
+            # So subtracting the tile size inverts the axis
+            pt = pt - self.tile_size
+            return pt if pt != 0 else None
+
+        x_px_end, y_px_end = invert_point(x_px_end), invert_point(y_px_end)
+        return image[y_px_start:y_px_end, x_px_start:x_px_end, :]
+
+    def _validate_num_tiles(self, xstart: float, ystart: float, xend: float,
+                            yend: float, max_tiles: int):
+        """Calculates the number of expected tiles we would fetch.
+
+        If this is greater than the number of max tiles, raise an error.
+        """
+        total_n_tiles = (yend - ystart + 1) * (xend - xstart + 1)
+        if total_n_tiles > max_tiles:
+            raise ValueError(f"Requested zoom results in {total_n_tiles} tiles."
+                             f"Max allowed tiles are {max_tiles}")
```

Comment:
Let's update this error message to tell the user that they can increase max tiles or reduce the zoom to solve the problem.